components/support/nimbus-fml/src/editing/error_path.rs

```diff
@@ -0,0 +1,101 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#[derive(Clone)]
+pub(crate) struct ErrorPath {
+    pub(crate) literals: Vec<String>,
+    pub(crate) path: String,
+}
+
+impl ErrorPath {
+    pub(crate) fn feature(name: &str) -> Self {
+        Self {
+            path: format!("features/{name}"),
+            literals: Default::default(),
+        }
+    }
+
+    pub(crate) fn object(name: &str) -> Self {
+        Self {
+            path: format!("object/{name}"),
+            literals: Default::default(),
+        }
+    }
+
+    pub(crate) fn property(&self, prop_key: &str) -> Self {
+        Self {
+            path: format!("{}.{prop_key}", &self.path),
+            literals: append_quoted(&self.literals, prop_key),
+        }
+    }
+
+    pub(crate) fn enum_map_key(&self, enum_: &str, key: &str) -> Self {
+        Self {
+            path: format!("{}[{enum_}#{key}]", &self.path),
+            literals: append_quoted(&self.literals, key),
+        }
+    }
+
+    pub(crate) fn map_key(&self, key: &str) -> Self {
+        Self {
+            path: format!("{}['{key}']", &self.path),
+            literals: append(&self.literals, &["{".to_string(), format!("\"{key}\"")]),
+        }
+    }
+
+    pub(crate) fn array_index(&self, index: usize) -> Self {
+        let mut literals = append1(&self.literals, "[");
+        for _ in 1..index {
+            literals.push(",".to_string());
+        }
+        Self {
+            path: format!("{}[{index}]", &self.path),
+            literals,
+        }
+    }
+
+    pub(crate) fn object_value(&self, name: &str) -> Self {
+        Self {
+            path: format!("{}#{name}", &self.path),
+            literals: append1(&self.literals, "{"),
+        }
+    }
+
+    pub(crate) fn open_brace(&self) -> Self {
+        Self {
+            path: self.path.clone(),
+            literals: append1(&self.literals, "{"),
+        }
+    }
+
+    pub(crate) fn final_error(&self, hightlight: &str) -> Self {
+        Self {
+            path: self.path.clone(),
+            literals: append1(&self.literals, hightlight),
+        }
+    }
+
+    pub(crate) fn final_error_quoted(&self, hightlight: &str) -> Self {
+        Self {
+            path: self.path.clone(),
+            literals: append_quoted(&self.literals, hightlight),
+        }
+    }
+}
+
+fn append(original: &[String], new: &[String]) -> Vec<String> {
+    let mut clone = original.to_owned();
+    clone.extend(new.iter().cloned());
+    clone
+}
+
+fn append1(original: &[String], new: &str) -> Vec<String> {
+    let mut clone = original.to_owned();
```

Comment:
If original.len is a power of two, then we will re-allocate when we push new.to_string(). We know how big the clone will be so we should reserve at least that much.

```suggestion
    let mut clone = Vec::with_capacity(original.len() + 1);
    clone.extend_from_slice(original);
    clone.push(new.to_string());
    clone
```